File: ios/Classes/full/YandexSearchSession.swift
import Foundation
import YandexMapsMobile

public class YandexSearchSession: NSObject {
  private var id: Int
  private var session: YMKSearchSession!
  private let methodChannel: FlutterMethodChannel!
  private let searchManager: YMKSearchManager
  private var page = 0
  private static var searchSessions: [Int: YandexSearchSession] = [:]

  public static func initSession(id: Int, registrar: FlutterPluginRegistrar, searchManager: YMKSearchManager) {
    searchSessions[id] = YandexSearchSession(id: id, registrar: registrar, searchManager: searchManager)
  }

  public required init(id: Int, registrar: FlutterPluginRegistrar, searchManager: YMKSearchManager) {
    self.id = id
    self.methodChannel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_search_session_\(id)",
      binaryMessenger: registrar.messenger()
    )
    self.searchManager = searchManager

    super.init()

    weak var weakSelf = self
    self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "searchByText":
      searchByText(call, result)
    case "searchByPoint":
      searchByPoint(call, result)
    case "cancel":
      cancel()
      result(nil)
    case "retry":
      retry(result)
    case "hasNextPage":
      let value = hasNextPage()
      result(value)
    case "fetchNextPage":
      fetchNextPage(result)
    case "close":
      close()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func searchByText(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]

    session = searchManager.submit(
      withText: params["searchText"] as! String,
      geometry: UtilsFull.geometryFromJson(params["geometry"] as! [String: Any]),
      searchOptions: UtilsFull.searchOptionsFromJson(params["searchOptions"] as! [String: Any]),
      responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
        self.handleResponse(searchResponse: searchResponse, error: error, result: result)
      }
    )
  }

  public func searchByPoint(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]

    session = searchManager.submit(
      with: UtilsFull.pointFromJson(params["point"] as! [String: NSNumber]),
      zoom: params["zoom"] as? NSNumber,
      searchOptions: UtilsFull.searchOptionsFromJson(params["searchOptions"] as! [String: Any]),
      responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
        self.handleResponse(searchResponse: searchResponse, error: error, result: result)
      }
    )
  }

  public func cancel() {
    session.cancel()
  }

  public func retry(_ result: @escaping FlutterResult) {
    page = 0

    session.retry(responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
      self.handleResponse(searchResponse: searchResponse, error: error, result: result)
    })
  }

  public func hasNextPage() -> Bool {
    return session.hasNextPage()
  }

  public func fetchNextPage(_ result: @escaping FlutterResult) {
    if (session.hasNextPage()) {
      page += 1

      session.fetchNextPage(responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
        self.handleResponse(searchResponse: searchResponse, error: error, result: result)
      })
    }
  }

  public func close() {
    session.cancel()

    YandexSearchSession.searchSessions.removeValue(forKey: id)
  }

  public func handleResponse(searchResponse: YMKSearchResponse?, error: Error?, result: @escaping FlutterResult) {
    if let response = searchResponse {
      onSuccess(response, result)
    } else {
      onError(error!, result)
    }
  }

  private func onSuccess(_ res: YMKSearchResponse, _ result: @escaping FlutterResult) {
    let items: [[String: Any?]] = res.collection.children.compactMap { item in
      let obj = item.obj

      if (obj == nil) {
        return nil
      }

      return [
        "name": obj!.name,
        "geometry": obj!.geometry.map { UtilsFull.geometryToJson($0) },
        "toponymMetadata": getToponymMetadata(metadataContainer: obj!.metadataContainer),
        "businessMetadata": getBusinessMetadata(metadataContainer: obj!.metadataContainer)
      ]
    }

    let arguments: [String: Any] = [
      "found": res.metadata.found,
      "page": page,
      "items": items
    ]

    result(arguments)
  }

  private func onError(_ error: Error, _ result: @escaping FlutterResult) {
    result(UtilsFull.errorToJson(error))
  }

  private func getToponymMetadata(metadataContainer: YRTTypeDictionary<YMKBaseMetadata>) -> [String: Any]? {
    let meta = metadataContainer.getItemOf(YMKSearchToponymObjectMetadata.self) as? YMKSearchToponymObjectMetadata

    if (meta == nil) {
      return nil
    }

    return [
      "balloonPoint": UtilsFull.pointToJson(meta!.balloonPoint),
      "address": [
        "formattedAddress": meta!.address.formattedAddress,
        "addressComponents": getAddressComponents(address: meta!.address)
      ]
    ]
  }

  private func getBusinessMetadata(metadataContainer: YRTTypeDictionary<YMKBaseMetadata>) -> [String: Any?]? {
    let meta = metadataContainer.getItemOf(YMKSearchBusinessObjectMetadata.self) as? YMKSearchBusinessObjectMetadata

    if (meta == nil) {
      return nil
    }

    return [
      "name": meta!.name,
      "shortName": meta!.shortName,
      "address": [
        "formattedAddress": meta!.address.formattedAddress,
        "addressComponents": getAddressComponents(address: meta!.address)
      ]
    ]
  }

  private func getAddressComponents(address: YMKSearchAddress) -> [Int: String] {
    var addressComponents = [Int : String]()

    address.components.forEach {
      let value = $0.name

      $0.kinds.forEach {
        var flutterKind: Int = 0
        let kind = YMKSearchComponentKind(rawValue: UInt(truncating: $0))

        switch kind {
        case .country:
          flutterKind = 1
        case .some(.region):
          flutterKind = 2
        case .some(.province):
          flutterKind = 3
        case .some(.area):
          flutterKind = 4
        case .some(.locality):
          flutterKind = 5
        case .some(.district):
          flutterKind = 6
        case .some(.street):
          flutterKind = 7
        case .some(.house):
          flutterKind = 8
        case .some(.entrance):
          flutterKind = 9
        case .some(.route):
          flutterKind = 10
        case .some(.station):
          flutterKind = 11
        case .some(.metroStation):
          flutterKind = 12
        case .some(.railwayStation):
          flutterKind = 13
        case .some(.vegetation):
          flutterKind = 14
        case .some(.hydro):
          flutterKind = 15
        case .some(.airport):
          flutterKind = 16
        case .some(.other):
          flutterKind = 17
        default:
          flutterKind = 0
        }

        addressComponents[flutterKind] = value
      }
    }

    return addressComponents
  }
}

File: android/src/main/java/com/unact/yandexmapkit/full/YandexSearchSession.java
package com.unact.yandexmapkit.full;

import androidx.annotation.NonNull;

import com.yandex.mapkit.BaseMetadata;
import com.yandex.mapkit.GeoObject;
import com.yandex.mapkit.GeoObjectCollection;
import com.yandex.mapkit.geometry.Geometry;
import com.yandex.mapkit.search.Address;
import com.yandex.mapkit.search.BusinessObjectMetadata;
import com.yandex.mapkit.search.Response;
import com.yandex.mapkit.search.SearchManager;
import com.yandex.mapkit.search.Session;
import com.yandex.mapkit.search.ToponymObjectMetadata;
import com.yandex.runtime.Error;
import com.yandex.runtime.TypeDictionary;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;

public class YandexSearchSession implements MethodChannel.MethodCallHandler {
  private final int id;
  private Session session;
  private final MethodChannel methodChannel;
  private final SearchManager searchManager;
  private int page = 0;
  @SuppressWarnings({"MismatchedQueryAndUpdateOfCollection"})
  private static final Map<Integer, YandexSearchSession> searchSessions = new HashMap<>();

  public static void initSession(int id, BinaryMessenger messenger, SearchManager searchManager) {
    searchSessions.put(id, new YandexSearchSession(id, messenger, searchManager));
  }

  public YandexSearchSession(
    int id,
    BinaryMessenger messenger,
    SearchManager searchManager
    ) {
    this.id = id;
    this.searchManager = searchManager;

    methodChannel = new MethodChannel(messenger, "yandex_mapkit/yandex_search_session_" + id);
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "searchByText":
        searchByText(call, result);
        break;
      case "searchByPoint":
        searchByPoint(call, result);
        break;
      case "cancel":
        cancel();
        result.success(null);
        break;
      case "retry":
        retry(result);
        break;
      case "hasNextPage":
        boolean value = hasNextPage();
        result.success(value);
        break;
      case "fetchNextPage":
        fetchNextPage(result);
        break;
      case "close":
        close();
        result.success(null);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void searchByText(MethodCall call, MethodChannel.Result result) {
    YandexSearchSession self = this;
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    session = searchManager.submit(
      (String) params.get("searchText"),
      UtilsFull.geometryFromJson((Map<String, Object>) params.get("geometry")),
      UtilsFull.searchOptionsFromJson((Map<String, Object>) params.get("searchOptions")),
      new Session.SearchListener() {
        @Override
        public void onSearchResponse(@NonNull Response response) { self.onSearchResponse(response, result); }
        @Override
        public void onSearchError(@NonNull Error error) { self.onSearchError(error, result); }
      }
    );
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void searchByPoint(MethodCall call, MethodChannel.Result result) {
    YandexSearchSession self = this;
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    session = searchManager.submit(
      UtilsFull.pointFromJson((Map<String, Object>) params.get("point")),
      ((Integer) params.get("zoom")),
      UtilsFull.searchOptionsFromJson((Map<String, Object>) params.get("searchOptions")),
      new Session.SearchListener() {
        @Override
        public void onSearchResponse(@NonNull Response response) { self.onSearchResponse(response, result); }
        @Override
        public void onSearchError(@NonNull Error error) { self.onSearchError(error, result); }
      }
    );
  }

  public void cancel() {
    session.cancel();
  }

  public void retry(MethodChannel.Result result) {
    YandexSearchSession self = this;
    page = 0;

    session.retry(
      new Session.SearchListener() {
        @Override
        public void onSearchResponse(@NonNull Response response) { self.onSearchResponse(response, result); }
        @Override
        public void onSearchError(@NonNull Error error) { self.onSearchError(error, result); }
      }
    );
  }

  public boolean hasNextPage() {
    return session.hasNextPage();
  }

  public void fetchNextPage(MethodChannel.Result result) {
    YandexSearchSession self = this;

    if (session.hasNextPage()) {
      page++;

      session.fetchNextPage(
        new Session.SearchListener() {
          @Override
          public void onSearchResponse(@NonNull Response response) { self.onSearchResponse(response, result); }
          @Override
          public void onSearchError(@NonNull Error error) { self.onSearchError(error, result); }
        }
      );
    }
  }

  public void close() {
    session.cancel();
    methodChannel.setMethodCallHandler(null);

    searchSessions.remove(id);
  }

  private void onSearchResponse(@NonNull Response response, @NonNull Result result) {
    Map<String, Object> arguments = new HashMap<>();
    List<Map<String, Object>> dataItems = new ArrayList<>();

    for (GeoObjectCollection.Item item : response.getCollection().getChildren()) {
      GeoObject obj = item.getObj();

      if (obj == null) {
        continue;
      }

      List<Map<String, Object>> geometryList = new ArrayList<>();
      for (Geometry geometry : obj.getGeometry()) {
        geometryList.add(UtilsFull.geometryToJson(geometry));
      }

      Map<String, Object> dataItem = new HashMap<>();
      dataItem.put("name", obj.getName());
      dataItem.put("geometry", geometryList);
      dataItem.put("toponymMetadata", getToponymMetadata(obj.getMetadataContainer()));
      dataItem.put("businessMetadata", getBusinessMetadata(obj.getMetadataContainer()));

      dataItems.add(dataItem);
    }

    arguments.put("found", response.getMetadata().getFound());
    arguments.put("page", page);
    arguments.put("items", dataItems);

    result.success(arguments);
  }

  private void onSearchError(@NonNull Error error, @NonNull Result result) {
    result.success(UtilsFull.errorToJson(error));
  }

  private Map<String, Object> getToponymMetadata(TypeDictionary<BaseMetadata> metadataContainer) {
    ToponymObjectMetadata meta = metadataContainer.getItem(ToponymObjectMetadata.class);

    if (meta == null) {
      return null;
    }

    Map<String, Object> address = new HashMap<>();
    address.put("formattedAddress", meta.getAddress().getFormattedAddress());
    address.put("addressComponents", getAddressComponents(meta.getAddress()));

    Map<String, Object> toponymMetadata = new HashMap<>();
    toponymMetadata.put("address", address);
    toponymMetadata.put("balloonPoint", UtilsFull.pointToJson(meta.getBalloonPoint()));

    return toponymMetadata;
  }

  private Map<String, Object> getBusinessMetadata(TypeDictionary<BaseMetadata> metadataContainer) {
    BusinessObjectMetadata meta = metadataContainer.getItem(BusinessObjectMetadata.class);

    if (meta == null) {
      return null;
    }

    Map<String, Object> address = new HashMap<>();
    address.put("formattedAddress", meta.getAddress().getFormattedAddress());
    address.put("addressComponents", getAddressComponents(meta.getAddress()));

    Map<String, Object> businessMetadata = new HashMap<>();
    businessMetadata.put("name", meta.getName());
    businessMetadata.put("shortName", meta.getShortName());
    businessMetadata.put("address", address);

    return businessMetadata;
  }

  private Map<Integer, String> getAddressComponents(Address address) {
    Map<Integer, String> addressComponents = new HashMap<>();

    for (Address.Component addressComponent : address.getComponents()) {
      String value = addressComponent.getName();

      for (Address.Component.Kind addressComponentKind : addressComponent.getKinds()) {
        int flutterKind = 0;

        switch (addressComponentKind) {
          case COUNTRY:
            flutterKind = 1;
            break;
          case REGION:
            flutterKind = 2;
            break;
          case PROVINCE:
            flutterKind = 3;
            break;
          case AREA:
            flutterKind = 4;
            break;
          case LOCALITY:
            flutterKind = 5;
            break;
          case DISTRICT:
            flutterKind = 6;
            break;
          case STREET:
            flutterKind = 7;
            break;
          case HOUSE:
            flutterKind = 8;
            break;
          case ENTRANCE:
            flutterKind = 9;
            break;
          case ROUTE:
            flutterKind = 10;
            break;
          case STATION:
            flutterKind = 11;
            break;
          case METRO_STATION:
            flutterKind = 12;
            break;
          case RAILWAY_STATION:
            flutterKind = 13;
            break;
          case VEGETATION:
            flutterKind = 14;
            break;
          case HYDRO:
            flutterKind = 15;
            break;
          case AIRPORT:
            flutterKind = 16;
            break;
          case OTHER:
            flutterKind = 17;
            break;
          default:
            break;
        }

        addressComponents.put(flutterKind, value);
      }
    }

    return addressComponents;
  }
}
